Password-protected Keystore Feature Branch PR

williamrandolph opened the pull request:
After months of back-and-forth, here's the PR for the password-protected keystore feature.

# Description

Elasticsearch uses a custom on-disk keystore for secure settings such as passwords and SSL certificates. Up until now, this prevented users with command-line access from viewing secure files by listing commands, but nothing prevented such users from changing values in the keystore, or removing values from it. Furthermore, the values were only obfuscated by a hash; no user-specific secret protected the secure settings.

This PR changes all of that by adding password-protection to the keystore. This will not be a breaking change for users. If a keystore has no password, there won’t be any new prompts. A user must choose to password-protect their keystore in order to see any new behavior.

This is a large feature branch, but the major pieces of production code are straightforward:

* The keystore CLI has been extended to allow users to create a password-protected keystore, to add password-protection to an existing keystore, and to change or remove a password from an existing keystore. Once a keystore has a password, the CLI will prompt for a password when listing, adding, or removing secure settings. There are new subcommands for the keystore CLI: “passwd” is for changing a keystore’s password, and “has-passwd” is for checking whether a keystore is password-protected. The common code for password handling in keystore CLI commands is contained in a new BaseKeystoreCommand class.
* The Elasticsearch startup scripts for Windows and Bash have been modified to read passwords on standard input. In the systemd case, a small entrypoint script can read the password from a file and pass it to the startup script. In the Docker case, the Docker entrypoint script can retrieve a password from an environment variable and provide it to the Elasticsearch startup script.
* The REST API command for reloading secure settings now requires a password when a node’s keystore is password-protected. A call to the reload_secure_settings with a password is allowed in two cases: (1) When run against the local node ( single node cluster, or specifically targeting the local cluster via _local filter or nodeId) regardless of the TLS configuration in the transport layer. (2) When run for more nodes than the local one (some or all cluster) iff TLS is enabled in the transport layer.
* The Command and Terminal classes have been modified to support finer-grained control over reading from standard input and writing errors to the terminal. Specifically, we read passwords into character arrays (wrapped as “secure strings”) up to a maximum length of 128 (arbitrarily chosen).
* Much of the new code in this PR is test code. Many of our already-existing keystore unit tests now use password-protected keystrokes. We also have extensive “qa” tests: a “KeystoreManagementTests” class that can run for both archive (zip, tar.gz), package (rpm, deb), and Docker distributions as part of the packaging test matrix.

There is one relevant bit of outstanding work: #49340. This feature branch includes two integration tests that use `expect` to run the Elasticsearch command with an interactive terminal, as opposed to merely piping input to a non-interactive shell. However, we have only been able to install `expect` on Linux test instances so far, so we don't yet have Windows parity. This work is blocked by an outstanding ticket for Infra.

Closes #32691

williamrandolph committed:
Password Protected Keystore (Feature Branch) (cleaned up) (#49268)

* Reload secure settings with password (#43197)

If a password is not set, we assume an empty string to be
compatible with previous behavior.
Only allow the reload to be broadcast to other nodes if TLS is
enabled for the transport layer.

* Add passphrase support to elasticsearch-keystore (#38498)

This change adds support for keystore passphrases to all subcommands
of the elasticsearch-keystore cli tool and adds a subcommand for
changing the passphrase of an existing keystore.
The work to read the passphrase in Elasticsearch when
loading, which will be addressed in a different PR.

Subcommands of elasticsearch-keystore can handle (open and create)
passphrase protected keystores

When reading a keystore, a user is only prompted for a passphrase
only if the keystore is passphrase protected.

When creating a keystore, a user is allowed (default behavior) to create one with an
empty passphrase

Passphrase can be set to be empty when changing/setting it for an
existing keystore

Relates to: #32691
Supersedes: #37472

* Restore behavior for force parameter (#44847)

Turns out that the behavior of `-f` for the add and add-file sub
commands where it would also forcibly create the keystore if it
didn't exist, was by design - although undocumented.
This change restores that behavior auto-creating a keystore that
is not password protected if the force flag is used. The force
OptionSpec is moved to the BaseKeyStoreCommand as we will presumably
want to maintain the same behavior in any other command that takes
a force option.

*  Handle pwd protected keystores in all CLI tools  (#45289)

This change ensures that `elasticsearch-setup-passwords` and
`elasticsearch-saml-metadata` can handle a password protected
elasticsearch.keystore.
For setup passwords the user would be prompted to add the
elasticsearch keystore password upon running the tool. There is no
option to pass the password as a parameter as we assume the user is
present in order to enter the desired passwords for the built-in
users.
For saml-metadata, we prompt for the keystore password at all times
even though we'd only need to read something from the keystore when
there is a signing or encryption configuration.

* Modify docs for setup passwords and saml metadata cli (#45797)

Adds a sentence in the documentation of `elasticsearch-setup-passwords`
and `elasticsearch-saml-metadata` to describe that users would be
prompted for the keystore's password when running these CLI tools,
when the keystore is password protected.

Co-Authored-By: Lisa Cawley <lcawley@elastic.co>

* Elasticsearch keystore passphrase for startup scripts (#44775)

This commit allows a user to provide a keystore password on Elasticsearch
startup, but only prompts when the keystore exists and is encrypted.

The entrypoint in Java code is standard input. When the Bootstrap class is
checking for secure keystore settings, it checks whether or not the keystore
is encrypted. If so, we read one line from standard input and use this as the
password. For simplicity's sake, we allow a maximum passphrase length of 128
characters. (This is an arbitrary limit and could be increased or eliminated.
It is also enforced in the keystore tools, so that a user can't create a
password that's too long to enter at startup.)

In order to provide a password on standard input, we have to account for four
different ways of starting Elasticsearch: the bash startup script, the Windows
batch startup script, systemd startup, and docker startup. We use wrapper
scripts to reduce systemd and docker to the bash case: in both cases, a
wrapper script can read a passphrase from the filesystem and pass it to the
bash script.

In order to simplify testing the need for a passphrase, I have added a
has-passwd command to the keystore tool. This command can run silently, and
exit with status 0 when the keystore has a password. It exits with status 1 if
the keystore doesn't exist or exists and is unencrypted.

A good deal of the code-change in this commit has to do with refactoring
packaging tests to cleanly use the same tests for both the "archive" and the
"package" cases. This required not only moving tests around, but also adding
some convenience methods for an abstraction layer over distribution-specific
commands.

I will write some user-facing documentation for these changes in a follow-up
commit.

* Adjust docs for password protected keystore (#45054)

This commit adds relevant parts in the elasticsearch-keystore
sub-commands reference docs and in the reload secure settings API
doc.

* Cleanup after feature branch reconstruction

The feature branch for the password-protected keystore, due to an
accident, contains a large number of unrelated commits. In order to get
a cleaner merge, I've cherry-picked the main commits that went into the
feature branch against a branch derived from master — essentially, a
rebase onto master. We've ignored some tests that will addressed in
follow-up PRs to the feature branch.

williamrandolph committed:
Merging master to feature branch

williamrandolph committed:
Merge branch 'master' into feature-pwd-protected-keystore-2

williamrandolph committed:
Fix failing Keystore Passphrase test for feature branch (#50154)

One problem with the passphrase-from-file tests, as written, is that
they would leave a SystemD environment variable set when they failed,
and this setting would cause elasticsearch startup to fail for other
tests as well. By using a try-finally, I hope that these tests will fail
more gracefully.

It appears that our Fedora and Ubuntu environments may be configured to
store journald information under /var rather than under /run, so that it
will persist between boots. Our destructive tests that read from the
journal need to account for this in order to avoid trying to limit the
output we check in tests.

williamrandolph committed:
Merge master into feature branch

williamrandolph committed:
Run keystore management tests on docker distros (#50610)

* Add Docker handling to PackagingTestCase

Keystore tests need to be able to run in the Docker case. We can do this
by using a DockerShell instead of a plain Shell when Docker is running.

* Improve ES startup check for docker

Previously we were checking truncated output for the packaged JDK as
an indication that Elasticsearch had started. With new preliminary
password checks, we might get a false positive from ES keystore
commands, so we have to check specifically that the Elasticsearch
class from the Bootstrap package is what's running.

williamrandolph committed:
Merging master into feature branch

williamrandolph committed:
Test password-protected keystore with Docker (#50803)

This commit adds two tests for the case where we mount a
password-protected keystore into a Docker container and provide a
password via a Docker environment variable.

We also fix a logging bug where we were logging the identifier for an
array of strings rather than the contents of that array.

williamrandolph committed:
Merging changes from master

williamrandolph committed:
Add documentation for keystore startup prompting (#50821)

When a keystore is password-protected, Elasticsearch will prompt at
startup. This commit adds documentation for this prompt for the archive,
systemd, and Docker cases.

Co-authored-by: Lisa Cawley <lcawley@elastic.co>

williamrandolph committed:
Merge branch 'master' into feature-pwd-protected-keystore-2

williamrandolph committed:
Warn when unable to upgrade keystore on debian (#51011)

For Red Hat RPM upgrades, we warn if we can't upgrade the keystore. This
commit brings the same logic to the code for Debian packages. See the
posttrans file for gets executed for RPMs.

williamrandolph committed:
Merge master into feature branch

williamrandolph committed:
Merge branch 'feature-pwd-protected-keystore-2' of github.com:elastic/elasticsearch into feature-pwd-protected-keystore-2

elasticmachine commented:
Pinging @elastic/es-core-infra (:Core/Infra/Packaging)

elasticmachine commented:
Pinging @elastic/es-security (:Security/Authentication)

williamrandolph committed:
Increase Docker startup wait time for tests

williamrandolph commented:
@elasticmachine run elasticsearch-ci/2

rjernst reviewed (approved):
I only did a cursory review since I've already looked at most of the PRs going into this branch, but this LGTM. Great work having all this behavior, tests and docs covered!

jkakavas reviewed (approved):
This LGTM ! However, as there are parts in here that I've written and parts that fall too much in the core-infra side ( docker, bash scripts ), I think it would make sense for another person from es-core-infra to make a pass as the second reviewer

pugnascotia commented on distribution/packages/src/common/scripts/postinst:
```diff
@@ -108,7 +108,12 @@ if [ "$PACKAGE" = "deb" ]; then
-        /usr/share/elasticsearch/bin/elasticsearch-keystore upgrade
+        if /usr/share/elasticsearch/bin/elasticsearch-keystore has-passwd --silent ; then
+          echo "### Warning: unable to upgrade encrypted keystore" 1>&2
+          echo " Please run elasticsearch-keystore upgrade and enter password" 1>&2
```
Tiny nit / observation - is it worth putting the command in quotes?

```suggestion
          echo " Please run 'elasticsearch-keystore upgrade' and enter password" 1>&2
```

pugnascotia commented on distribution/src/bin/elasticsearch:
```diff
@@ -20,6 +20,19 @@ if [ -z "$ES_TMPDIR" ]; then
+# conflicting GC configurations for the keystore tools
+unset KEYSTORE_PASSWORD
+KEYSTORE_PASSWORD=
+if ! echo $* | grep -E -q '(^-h |-h$| -h |--help$|--help |^-V |-V$| -V |--version$|--version )' \
```
This feels a bit fragile to me. What about something like:

```shell
CHECK_KEYSTORE=false
for option in "$@"; do
  case "$option" in
    -h|--help|-V|--version)
      CHECK_KEYSTORE=true
      ;;
  esac
done
if [[ $CHECK_KEYSTORE == true ]] \
    && "`dirname "$0"`"/elasticsearch-keystore has-passwd --silent
  # etc
```

williamrandolph replied:
I copied a pattern checking for daemonization below. I'd like to put this off for a follow-up PR so it can be discussed with some other team members in isolation. That being said I don't have any objection to doing a loop over options and it seems to work just fine in my quick local tests.

pugnascotia commented on distribution/tools/keystore-cli/src/main/java/org/elasticsearch/common/settings/CreateKeyStoreCommand.java:
```diff
@@ -21,41 +21,44 @@
+            readPassword(terminal, true) : new SecureString(new char[0])) {
+            Path keystoreFile = KeyStoreWrapper.keystorePath(env.configFile());
+            if (Files.exists(keystoreFile)) {
+                if (terminal.promptYesNo("An elasticsearch keystore already exists. Overwrite?", false) == false) {
```
Should be we capitalizing "elasticsearch" here, and elsewhere?

```suggestion
                if (terminal.promptYesNo("An Elasticsearch keystore already exists. Overwrite?", false) == false) {
```

jkakavas replied:
I think we use non-capitalized `elasticsearch` everywhere else in our logs, prompts ,etc.

pugnascotia commented on distribution/tools/keystore-cli/src/test/java/org/elasticsearch/common/settings/AddFileKeyStoreCommandTests.java:
```diff
@@ -53,110 +53,156 @@ private Path createRandomFile() throws IOException {
+        assertSecureFile("foo", file2, password);
     }
 
     public void testForceNonExistent() throws Exception {
```
Nit-pick - I first read this test name as meaning that the file being added doesn't exist. Could we tweak it, e.g.

```suggestion
    public void testForceDoesNotAlreadyExist() throws Exception {
```

More generally, it'd be great if these tests had Javadoc comments that briefly explained each scenario / expectation, but I'll leave that up to you.

jkakavas replied:
I think this is a generic problem for a lot of our tests. It's not quite obvious all the time what the test does and the method name can only have so many words in it. However, very few of our tests have javadocs. 

In this case I think we can get away with editing the method names but feel free to argue against it

pugnascotia commented on distribution/tools/keystore-cli/src/test/java/org/elasticsearch/common/settings/AddStringKeyStoreCommandTests.java:
```diff
@@ -70,105 +80,92 @@ public void testMissingNoCreate() throws Exception {
-        assertSecureString("foo", "Typedthisandhitenter");
-    }
-
-    public void testAddUtf8String() throws Exception {
```
What happened to these tests?

jkakavas replied:
These were removed in https://github.com/elastic/elasticsearch/pull/38498 but apparently by mistake, I don't remember removing them for a reason. I'll push them back in

jkakavas replied:
Good catch @pugnascotia ! Not only these were removed by mistake, but the removal was hiding a bug :pray:

pugnascotia commented on distribution/tools/keystore-cli/src/test/java/org/elasticsearch/common/settings/HasPasswordKeyStoreCommandTests.java:
```diff
@@ -0,0 +1,68 @@
+
+    public void testFailsWithNoKeystore() throws Exception {
+        UserException e = expectThrows(UserException.class, this::execute);
+        assertEquals("Exit code should be 1", HasPasswordKeyStoreCommand.NO_PASSWORD_EXIT_CODE, e.exitCode);
```
Maybe not hard-code the expected exit code in the message? Very nit-picky, I realise.

```suggestion
        assertEquals("Unexpected exit code", HasPasswordKeyStoreCommand.NO_PASSWORD_EXIT_CODE, e.exitCode);
```

lcawl commented on docs/reference/cluster/nodes-reload-secure-settings.asciidoc:
```diff
@@ -1,13 +1,11 @@
 [[cluster-nodes-reload-secure-settings]]
-== Nodes Reload Secure Settings
+=== Nodes reload secure settings API
+++++
```
This API page should follow the format specified in https://github.com/elastic/docs/blob/master/shared/api-ref-ex.asciidoc (similar to the other cluster APIs).  This cleanup can occur in a subsequent PR, however.

pugnascotia commented on qa/os/src/test/java/org/elasticsearch/packaging/test/KeystoreManagementTests.java:
```diff
@@ -0,0 +1,416 @@
+
+public class KeystoreManagementTests extends PackagingTestCase {
+
+    private static final String PASSWORD_ERROR_MESSAGE = "Provided keystore password was incorrect";
```
Should we extract some of the other standard error messages too?

pugnascotia commented on qa/os/src/test/java/org/elasticsearch/packaging/test/KeystoreManagementTests.java:
```diff
@@ -0,0 +1,416 @@
+        Map<String, String> envVars = Map.of("KEYSTORE_PASSWORD", password);
+        runContainer(distribution(), volumes, envVars);
+        waitForElasticsearch(installation);
+        ServerUtils.runElasticsearchTests();
```
Is it not sufficient to just ensure that Elasticsearch has started? Do we need to also run the tests?

williamrandolph replied:
I don't see the downside in checking that Elasticsearch is not just running but minimally functional as well.

pugnascotia commented on server/src/main/java/org/elasticsearch/cli/KeyStoreAwareCommand.java:
```diff
@@ -0,0 +1,86 @@
+            if (Arrays.equals(passwordArray, passwordVerification) == false) {
+                throw new UserException(ExitCodes.DATA_ERROR, "Passwords are not equal, exiting.");
+            }
+            Arrays.fill(passwordVerification, '\u0000');
```
Any particular reason we're not using `'\0'` here, like we do in other places that we zero out an array?

jkakavas replied:
No reason, both are the null character denoted as a decimal or a hexadecimal. I think I have introduced all the `\u0000` instances as this is a convention I had been using

pugnascotia commented on server/src/test/java/org/elasticsearch/action/admin/ReloadSecureSettingsIT.java:
```diff
@@ -109,83 +115,213 @@ public void testInvalidKeystoreFile() throws Exception {
+                    try {
+                        assertThat(nodesReloadResponse, notNullValue());
+                        final Map<String, NodesReloadSecureSettingsResponse.NodeResponse> nodesMap = nodesReloadResponse.getNodesMap();
+                        assertThat(nodesMap.size(), equalTo(1));
```
Tiny nit: you can use `hasSize`.

```suggestion
                        assertThat(nodesMap, hasSize(1));
```

jkakavas replied:
Actually there is no hasSize for  `Matcher<java.util.Map<? extends K,? extends V>>`

pugnascotia replied:
Ah, there's `aMapWithSize()`. Don't know if that's an improvement 🤷‍♂

jkakavas committed:
Restore handling of string input

Adds tests that were mistakenly removed. One of these tests proved
we were not handling the the stdin (-x) option correctly when no
input was added. This commit restores the original approach of
reading stdin one char at a time until there is no more (-1, \r, \n)
instead of using readline() that might return null

jkakavas committed:
address feedback

williamrandolph committed:
Apply spotless reformatting

williamrandolph committed:
Merging changes from master

williamrandolph committed:
Reformat after merge from master

williamrandolph committed:
Back out merge commit errors

williamrandolph committed:
Respond to PR comments

williamrandolph committed:
Stop journald before trying to delete its files

williamrandolph committed:
Adjust clearing the journalctl logs

williamrandolph committed:
Merging changes from master

williamrandolph committed:
Use '--since' flag to get recent journal messages

When we get Elasticsearch logs from journald, we want to fetch only log
messages from the last run. There are two reasons for this. First, if
there are many logs, we might get a string that's too large for our
utility methods. Second, when we're looking for a specific message or
error, we almost certainly want to look only at messages from the last
execution.

Previously, we've been trying to do this by clearing out the physical
files under the journald process. But there seems to be some contention
over these directories: if journald writes a log file in between when
our deletion command deletes the file and when it deletes the log
directory, the deletion will fail.

It seems to me that we might be able to use journald's "--since" flag to
retrieve only log messages from the last run, and that this might be
less likely to fail due to race conditions in file deletion.

Unfortunately, it looks as if the "--since" flag has a granularity of
one-second. I've added a two-second sleep to make sure that there's a
sufficient gap between the test that will read from journald and the
test before it.

williamrandolph commented:
@elasticmachine run elasticsearch-ci/1

@elasticmachine run elasticsearch-ci/2

williamrandolph committed:
Use --since when reading journal

williamrandolph committed:
Merging changes from master

williamrandolph committed:
Use new journald wrapper pattern

williamrandolph committed:
Update version added in secure settings request
